<a href="https://colab.research.google.com/github/bjungweapon/mjc.ai.ml/blob/BDU/BDU_5_3_PCA_%EC%B0%A8%EC%9B%90%EC%B6%95%EC%86%8C_iris.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

데이터 전처리 - 차원 축소  중
PCA (자원 축소)

여러 변수들의 변량을 주성분이라고 한다. **서로 상관성이 높은 여러 변수들을  새로운 변수들로 요약 , 축소하는 기법. **

다변량 통계 분석에 적용한다.

iris의 데이터 를 이용하여 꽃잎과 꽃 받침의 길이와 너비를 이용하여 꽃의 종류를 구분(분류)할 수 있는가에 대한 처리

In [19]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [20]:
## 측정치 (관측치)를 모은 자료

iris = pd.read_csv ("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv")
iris.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [23]:
iris

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

다섯가지 종류와 , data 종류 150개, 꽃잎과 너비와 길이, 꽃 받침의 너비와 길이가 주어진다.

## 연속형 변수와 범주형 변수를 분리하자.
 종류값은 범주형 변수이므로, 정수형 변수들과 결과를 분리하자.

In [28]:
 df=iris.drop(["species"],axis=1)  ## axis = 1 : 행 ,  drop : 삭제.
 df


sepal_length  sepal_width  petal_length  petal_width
0             5.1          3.5           1.4          0.2
1             4.9          3.0           1.4          0.2
2             4.7          3.2           1.3          0.2
3             4.6          3.1           1.5          0.2
4             5.0          3.6           1.4          0.2
..            ...          ...           ...          ...
145           6.7          3.0           5.2          2.3
146           6.3          2.5           5.0          1.9
147           6.5          3.0           5.2          2.0
148           6.2          3.4           5.4          2.3
149           5.9          3.0           5.1          1.8

[150 rows x 4 columns]

In [29]:
 df_species = iris["species"]
 df_species

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: species, Length: 150, dtype: object

종류와 성분들 간의 data가 분리되었다.

변수간의 스케일 차이가 나면 스케일이 큰 변수가 주성분에 영향을 많이 주기 때문에 주성분 분석 전에 변수를 표준화나 정규화 시켜줌

수치형 변수 정규화
 변수를 정규화 하면 평균0, 분산 1인 새로운 변수로 변환하게 된다. **굵은 텍스트** **굵은 텍스트**

In [41]:
from sklearn.preprocessing import StandardScaler  ## Standard Scaler

x = StandardScaler().fit_transform(df)
df["sepal_length"] = StandardScaler().fit_transform(df[["sepal_length"]])
df["sepal_width"] = StandardScaler().fit_transform(df[["sepal_width"]])
df["petal_length"] = StandardScaler().fit_transform(df[["petal_length"]])
df["petal_width"] = StandardScaler().fit_transform(df[["petal_width"]])
df

sepal_length  sepal_width  petal_length  petal_width
0       -0.900681     1.019004     -1.340227    -1.315444
1       -1.143017    -0.131979     -1.340227    -1.315444
2       -1.385353     0.328414     -1.397064    -1.315444
3       -1.506521     0.098217     -1.283389    -1.315444
4       -1.021849     1.249201     -1.340227    -1.315444
..            ...          ...           ...          ...
145      1.038005    -0.131979      0.819596     1.448832
146      0.553333    -1.282963      0.705921     0.922303
147      0.795669    -0.131979      0.819596     1.053935
148      0.432165     0.788808      0.933271     1.448832
149      0.068662    -0.131979      0.762758     0.790671

[150 rows x 4 columns]

In [40]:
## PCA 수행

from sklearn.decomposition import PCA

pca = PCA(n_components=4)   # n_componenets 는 변환할 차원의 수
p_score = pca.fit_transform(df)  ## score로 만들어 줌
print(p_score.shape)
print(pca.explained_variance_ratio_)  ##변환된 각각의 차원의 기여도 (상관성)

(150, 4)
[0.72962445 0.22850762 0.03668922 0.00517871]


**기여도**

72.9%  : 73%로 설명이 가능하다. 꽃핑의 길이 만으로도 전체 data의 변동성을 설명할 수 있다.
22.8%  :
3.6 %
0.5%
즉  앞의 두 변수 만으로도 기여도를 봤을 때, Sepal_Length와 Sepal_Width 만으로도 분류가 가능하다는 것을 알 수 있다.